presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,43 +611,202 @@ private static NestedField nestedField(String name)
+                            context.setLambdaSubfields(lambdaSubfields.get(callArgumentIndex));
+                        }
+                        else {
+                            context.setLambdaSubfields(Context.allSubfieldsOfArrayElementOrMapValue);
+                        }
+                        call.getArguments().get(callArgumentIndex).accept(this, context);
+                    }
+
+                    // When we are done with inner calls (child nodes) we need to restore lambda subfields we received from parent expression to handle such situations like
+                    // in example below
+                    // SELECT * FROM my_table WHERE ANY_MATCH(column1, x -> x.ds > '2023-01-01') AND ALL_MATCH(column2, x -> STRPOS(x.comment,  'Presto') > 0)
+                    // After we are done with ANY_MATCH, we need to restore the lambda subfields to what we received from parent node 'AND' so that it does not collide with
+                    // lambda subfields of ALL_MATCH function.
+                    context.setLambdaSubfields(lambdaSubfieldsOriginal);
+                    return null;
                 }
+                context.setLambdaSubfields(Context.allSubfieldsOfArrayElementOrMapValue);
+                call.getArguments().forEach(argument -> argument.accept(this, context));
                 return null;
             }
 
+            private static Map<Integer, Set<Subfield>> merge(Map<Integer, Set<Subfield>> s1, Map<Integer, Set<Subfield>> s2)
+            {
+                Map<Integer, Set<Subfield>> result = new HashMap<>(s1);
+                s2.forEach((callArgumentIndex, subfields) -> result.merge(
+                        callArgumentIndex,
+                        subfields,
+                        (lambdaSubfields1, lambdaSubfields2) -> ImmutableSet.<Subfield>builder().addAll(lambdaSubfields1).addAll(lambdaSubfields2).build()));
+                return ImmutableMap.copyOf(result);
+            }
+
+            private static Map<Integer, Set<Subfield>> addNoSubfieldIfNoAccessedSubfieldsFound(CallExpression call, Map<Integer, Set<Subfield>> argumentIndexToLambdaSubfieldsMap)
+            {
+                ImmutableMap.Builder<Integer, Set<Subfield>> argumentIndexToLambdaSubfieldsMapBuilder = ImmutableMap.builder();
+                for (Integer callArgumentIndex : argumentIndexToLambdaSubfieldsMap.keySet()) {
+                    if (!argumentIndexToLambdaSubfieldsMap.get(callArgumentIndex).isEmpty()) {
+                        argumentIndexToLambdaSubfieldsMapBuilder.put(callArgumentIndex, argumentIndexToLambdaSubfieldsMap.get(callArgumentIndex));
+                    }
+                    else {
+                        RowExpression argument = call.getArguments().get(callArgumentIndex);
+                        if ((argument.getType() instanceof ArrayType && ((ArrayType) argument.getType()).getElementType() instanceof RowType) ||
```

Comment:
What about 2d arrays, or cases like array<map<...>>?

Reply:
This is a very great catch! Currently, we support only `NoSubfield` (e.g. `.$`) which is the only for Row type. We can extend it and add `NoSubscript` (e.g. `[$]`)  indicating that we are not interested in any subscript of the nested array or map.

Reply:
That's worth a comment

Reply:
Created the issue #19428 for that